src/provider/provider_tracking.c

```diff
@@ -216,6 +260,116 @@ static const char *trackingName(void *provider) {
     return umfMemoryProviderGetName(p->hUpstream);
 }
 
+static enum umf_result_t trackingGetIpcHandleSize(void *provider,
+                                                  size_t *size) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+    return umfMemoryProviderGetIPCHandleSize(p->hUpstream, size);
+}
+
+static enum umf_result_t trackingGetIpcHandle(void *provider, const void *ptr,
+                                              size_t size, void *ipcData) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+    enum umf_result_t ret = UMF_RESULT_SUCCESS;
+    size_t ipcDataSize = 0;
+    int cached = 0;
+    do {
+        void *value = critnib_get(p->ipcCache, (uintptr_t)ptr);
+        if (value) { //cache hit
+            struct ipc_cache_value_t *cache_value =
+                (struct ipc_cache_value_t *)value;
+            memcpy(ipcData, cache_value->ipcData, cache_value->size);
+            cached = 1;
+        } else {
+            ret =
+                umfMemoryProviderGetIPCHandle(p->hUpstream, ptr, size, ipcData);
+            if (ret != UMF_RESULT_SUCCESS) {
+                return ret;
+            }
+
+            ret = umfMemoryProviderGetIPCHandleSize(p->hUpstream, &ipcDataSize);
+            assert(ret == UMF_RESULT_SUCCESS);
+
+            size_t value_size = sizeof(struct ipc_cache_value_t) + ipcDataSize;
+            struct ipc_cache_value_t *cache_value =
+                (struct ipc_cache_value_t *)malloc(value_size);
+
+            cache_value->size = ipcDataSize;
+            memcpy(cache_value->ipcData, ipcData, ipcDataSize);
+
+            int insRes = critnib_insert(p->ipcCache, (uintptr_t)ptr,
+                                        (void *)cache_value, 0 /*update*/);
+            if (insRes == 0) {
+                cached = 1;
+            } else {
+                // critnib_insert might fail in 2 cases:
+                // 1. Another thread created cache entry. So we need to
```

Comment:
Why can't we just return the `ipcData` we have even if another thread inserted into the cache? We will then have two different IPC handles to the same memory but is it a problem (for correctness or performance)?

Reply:
Current implementation of put function does nothing today. The implementation rely on ipcCache and we close ipc handle when free method is called.
I considered the same idea like you suggested but it requires put method to be thread-safe and check whether ipcData is in the cache or not.

Reply:
Oh, right, because we might have a leak otherwise.